File: PAT/basic_level/1013_数素数/prime_table.cpp
#include <iostream>
#include <cmath>
#define MAX 150001
using namespace std;

int main(){
	int i=0, j=0;
	int up_bound=0;
	int us_pm=0, us_pn=0, cnt=0, line_cnt=0;
	unsigned long int num=2;

	cin>>us_pm>>us_pn;

	bool prime_list[MAX]={false, false, true};
	for(i=3; i<MAX; i+=2){
		prime_list[i] = true;
		prime_list[i+1] = false;
	}

	up_bound = (int)sqrt(MAX);
	for(i=3; i<=up_bound; i+=2){
		if(prime_list[i])
			for(j=i+i; j<MAX; j+=i)
				prime_list[j]=false;
	}

	while(1){
		if(prime_list[num]){
			cnt++;
		}
		if(us_pm == cnt){
			cout<<num;
			line_cnt++;
			num++;
			break;
		}
		num++;
	}

	if (us_pm >= us_pn){
		return 0;
	}

	while(1){
		if(prime_list[num]){
			if (9 == line_cnt){
				cout<<' '<<num<<endl;
				line_cnt = 0;
			}else{
				(line_cnt)?cout<<' '<<num:cout<<num;
				line_cnt++;
			}
			cnt++;
			if(us_pn == cnt){
				break;
			}
		}
		num++;
	}
	return 0;
}

File: PAT/basic_level/1013_数素数/prime_table.py
#!/usr/bin/env python
from __future__ import print_function

us_pm, us_pn = raw_input().split()
us_pm, us_pn = int(us_pm), int(us_pn)
num, cnt, line_cnt = 2, 0, 0

MAX = us_pn*10 + 5001
is_prime = [bool] * MAX
is_prime[0:3] = [False, False, True]
for i in range(3, MAX, 2):
	is_prime[i] = True
	is_prime[i+1] = False

up_bound = int(MAX**0.5) + 1
for i in range(3, up_bound, 2):
	if is_prime[i]:
		for j in range(i+i, MAX, i):
			is_prime[j] = False

while True:
	if is_prime[num]:
		cnt += 1
	if us_pm == cnt:
		print(num, end='')
		line_cnt += 1
		num += 1
		break
	num += 1

if us_pn == us_pm:
	exit()

while True:
	if is_prime[num]:
		if 9 == line_cnt:
			print(' %d'%num)
			line_cnt = 0
		else:
			if 0==line_cnt:
				print('%d'%num, end='')
			else:
				print(' %d'%num, end='')
			line_cnt += 1
		cnt += 1
		if us_pn == cnt:
			break
	num += 1
